src/main/java/com/ldtteam/structurize/client/gui/WindowShapeTool.java

```diff
@@ -189,10 +190,13 @@ else if (pos != null)
         registerButton(BUTTON_ROTATE_LEFT, this::rotateLeftClicked);
         registerButton(BUTTON_PICK_MAIN_BLOCK, this::pickMainBlock);
         registerButton(BUTTON_PICK_FILL_BLOCK, this::pickFillBlock);
+        registerButton(BUTTON_UNDOREDO, b -> {
+            close();
+            new WindowUndoRedo().open();
+        });
+        PaneBuilders.tooltipBuilder().hoverPane(findPaneOfTypeByID(BUTTON_UNDOREDO, Button.class)).append(new TranslationTextComponent("structurize.gui.undoredo")).build();
```

Comment:
iirc you can do that from xml if it's single line translation

Reply:
fixed